pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -199,6 +245,27 @@ public void acknowledgeMessage(List<Position> positions, AckType ackType, Map<St
             if (log.isDebugEnabled()) {
                 log.debug("[{}][{}] Individual acks on {}", topicName, subName, positions);
             }
+
+            synchronized (PersistentSubscription.this) {
+                positions.forEach(position -> {
+                    // If single ack try to ack message in pending_ack status, fail the ack.
+                    if (this.pendingAckMessages.contains(position)) {
+                        log.warn("[{}][{}] Invalid acks position conflict with an ongoing transaction:{}{}.",
+                                topicName, subName, this.txnID.getMostSigBits(), this.txnID.getLeastSigBits());
+                        return;
+                    }
+
+                    // If single ack is within range of cumulative ack of an ongoing transaction, fail the ack.
+                    if (null != this.pendingCumulativeAckMessage &&
+                            ((PositionImpl) position).compareTo((PositionImpl) this.pendingCumulativeAckMessage) < 0) {
```

Comment:
should it be `<=`?

Reply:
yes, markDelete include the last position, so should be <=.